api/v1alpha1/ndb_api_helpers.go

```diff
@@ -155,43 +157,147 @@ func GetNoneTimeMachineSLA(ctx context.Context, ndbclient *ndbclient.NDBClient)
+}
+
+func (inputProfile *Profile) Resolve(ctx context.Context, allProfiles []ProfileResponse, pType string, filter func(p ProfileResponse) bool) (profile ProfileResponse, err error) {
+	log := ctrllog.FromContext(ctx)
+	log.Info("Entered ndb_api_helpers.resolve", "input profile", inputProfile)
+
+	name, id := inputProfile.Name, inputProfile.Id
+	isNameProvided, isIdProvided := name != "", id != ""
+
+	var profileByName, profileById ProfileResponse
+
+	// resolve the profile based on provided input, return an error if not resolved
+	if isNameProvided {
+		profileByName, err = util.FindFirst(allProfiles, func(p ProfileResponse) bool { return p.Name == name })
+	}
 
-	profiles, err := GetAllProfiles(ctx, ndbclient)
+	if isIdProvided && err == nil {
+		profileById, err = util.FindFirst(allProfiles, func(p ProfileResponse) bool { return p.Id == id })
+	}
 
 	if err != nil {
-		return
+		log.Error(err, "could not resolve profile by id or name", "profile type", pType, "id", id, "name", name)
+		return ProfileResponse{}, fmt.Errorf("could not resolve profile by id=%v or name=%v", id, name)
 	}
 
-	genericProfiles := util.Filter(profiles, func(p ProfileResponse) bool { return p.EngineType == DATABASE_ENGINE_TYPE_GENERIC })
-	dbEngineSpecificProfiles := util.Filter(profiles, func(p ProfileResponse) bool { return p.EngineType == GetDatabaseEngineName(dbType) })
-
-	// Specifying the usage of small compute profiles
-	computeProfiles := util.Filter(genericProfiles, func(p ProfileResponse) bool {
-		return p.Type == PROFILE_TYPE_COMPUTE && strings.Contains(strings.ToLower(p.Name), "small")
-	})
-	storageProfiles := util.Filter(genericProfiles, func(p ProfileResponse) bool { return p.Type == PROFILE_TYPE_STORAGE })
-	// Specifying the usage of single instance topoplogy
-	softwareProfiles := util.Filter(dbEngineSpecificProfiles, func(p ProfileResponse) bool { return p.Type == PROFILE_TYPE_SOFTWARE && p.Topology == TOPOLOGY_SINGLE })
-	networkProfiles := util.Filter(dbEngineSpecificProfiles, func(p ProfileResponse) bool { return p.Type == PROFILE_TYPE_NETWORK })
-	dbParamProfiles := util.Filter(dbEngineSpecificProfiles, func(p ProfileResponse) bool { return p.Type == PROFILE_TYPE_DATABASE_PARAMETER })
-
-	// Some profile not found, return an error
-	if len(computeProfiles) == 0 || len(softwareProfiles) == 0 || len(storageProfiles) == 0 || len(networkProfiles) == 0 || len(dbParamProfiles) == 0 {
-		err = errors.New("oob profile: one or more OOB profile(s) were not found")
+	/*
+		1. if both name & id not provided => resolve the OOB profile
+		2. else if both name & id are provided => resolve by both & ensure that both resolved profiles are match
+		3. else if only id provided => resolve by id
+		4. else if only name provided => resolve by name
+		5. else => throw an error
+	*/
+	if !isNameProvided && !isIdProvided { // OOB
```

Comment:
Would it make sense to move this check before fetching profileByName and  profileById and 176 & 180

Reply:
We recently re-structured this whole piece of code for more readability and re-use. Since util.FindFirst is not a DB/ Network call and simply a filtering logic, I don't think it may give us significant performance benefit. The current structure clubs together all conditions together to put things into perspective what all cases can exist. Let me know what do you think on this?

Reply:
Sure! We can discuss it offline.